### DISCUSSION TITLE
How to run multiple dbt commands in sequential order

### DISCUSSION THREAD

**Discussion by ajs-rl** [dagster-io/dagster#19191] [Q&A] [answered]
I have a series of dbt commands (various `build` and `run-operation` calls) that I want to run in a specific order.  My first thought was to do this using the `deps` parameter, but the `@dbt_assets` decorator doesn't support them.

I then found that it is possible to include multiple cli invocations in one dbt assets definition.  However, when doing so, it looks like the Dagster behavior is overriding my selection criteria and running all models in my project.

I am wondering 1) if there is a way to use the `deps` parameter to enforce dependencies between different dbt cli invocations and/or 2) how to allow my cli invocations to run as-is without the dbt selection override?

I think the issue in part is that I am using `build_schedule_from_dbt_selection` - which allows me to pass my collection of dbt assets that expects a selection whereas I am trying to control it more finely in my python function.

Build schedule:
```python
schedules = [
    build_schedule_from_dbt_selection(
        [my_dbt_project_dbt_assets],
        job_name="my_dbt_job",
        cron_schedule="*/20 * * * *",
        # dbt_select="fqn:*",  # tried with and without this active 
    ),
]
```

dbt assets:
```python 
@dbt_assets(   
    manifest=dbt_manifest_path
)
def my_dbt_project_dbt_assets(context: AssetExecutionContext, dbt: DbtCliResource):
        
    cli_commands = [

        {'name': 'first_models',
         'cli_args': [
             "build"
            ,"--select"
            ,"models/staging"
            ]
        },

        {'name': 'my_macro',
         'cli_args': [
            "run-operation"
            ,"my_macro"
            ]
        },

        {'name': 'second_models',
         'cli_args': [
            "build"
            ,"--select"
            ,"models/intermediate/+"
            ]
        }

...
]

    for cli_cmd in cli_commands:
        context.log.info(f'Now running command set <{cli_cmd["name"]}> with args {cli_cmd["cli_args"]}')
        yield from dbt.cli(cli_cmd['cli_args'], context=context).stream()
```

When running the job, I see just the first command being invoked:

```
Now running command set first_models with args ['build', '--select', 'models/staging']
Running dbt command: `dbt build --select models/staging --select fqn:*`.
```

The first line being my statement and the second being a default one.
I note that when I run this statement within dbt directly, it does not accept the syntax:
`dbt build --select models/staging --select fqn:*`

**Comment by HAMZA310** [comment, marked as answer]
> how to allow my cli invocations to run as-is without the dbt selection override?


I found a _workaround_ that, while not ideal, might help. Instead of passing the `context` in your dbt CLI invocation, try using only the `manifest`:
```Python
yield from dbt.cli(cli_cmd['cli_args'], manifest=dbt_manifest_path).stream()
```

I resorted to this workaround after exploring the Dagster documentation and not finding a specific solution to this issue.

---

P.S. [Based on the source code](https://github.com/dagster-io/dagster/blob/5f17893376f5bc1be2c19d7835e49f4f5d9eaf87/python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py#L948-L990), it appears that even when the dbt CLI is invoked with custom select arguments (e.g. `'cli_args': [
             "build"
            ,"--select"
            ,"models/staging"
            ]
`), it is still interpreted as a _non_-subsetted execution. Consequently, it appends the default dbt selection arguments (fqn:*), which take precedence.

**Comment by mjkanji** [comment]
@ajs-rl If you simply want to break up the `dbt build` invocations of different subsets of models, what you can do instead is to have multiple `@dbt_assets` decorated functions. I mostly use AMPs so I don't have to schedule jobs, but if you want to use jobs and schedules, then you just have to set up the correct sensors to trigger the first job on a schedule and the subsequent jobs after each other.

Here's an example of what I do:

```python
@dbt_assets(
    manifest=Path(dbt_manifest_path),
    io_manager_key="dbt_io_manager",
    dagster_dbt_translator=DailyDagsterDbtTranslator(),
    exclude="resource_type:snapshot",
)
def dbt_models(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_commands = (["run"], ["test"])

    for dbt_command in dbt_commands:
        yield from dbt.cli(dbt_command, context=context).stream()


@dbt_assets(
    manifest=Path(dbt_manifest_path),
    io_manager_key="dbt_io_manager",
    dagster_dbt_translator=DailyDagsterDbtTranslator(),
    select="tag:daily_snaps",
)
def daily_snapshots(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_commands = (["snapshot"],)

    for dbt_command in dbt_commands:
        yield from dbt.cli(dbt_command, context=context).stream()


@dbt_assets(
    manifest=Path(dbt_manifest_path),
    io_manager_key="dbt_io_manager",
    dagster_dbt_translator=WeeklyDagsterDbtTranslator(),
    select="tag:weekly_snaps",
)
def weekly_snapshots(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_commands = (["snapshot"],)

    for dbt_command in dbt_commands:
        yield from dbt.cli(dbt_command, context=context).stream()
```

The job definitions:

```python
# Job definitions
dbt_daily_refresh = define_asset_job(
    "dbt_daily_refresh",
    selection=[dbt_models, daily_snapshots]
)

dbt_weekly_refresh = define_asset_job(
    "dbt_weekly_refresh",
    selection=[dbt_models, daily_snapshots, weekly_snapshots]
)

# Schedules
dbt_daily_schedule = ScheduleDefinition(
    job=dbt_daily_refresh,
    cron_schedule="0 1 * * 0,2-6",
    description="Runs daily at 1 AM, except on Mondays. Used for Daily DBT refreshes.",
)

dbt_weekly_schedule = ScheduleDefinition(
    job=dbt_weekly_refresh,
    cron_schedule="0 1 * * 1",
    description="Runs daily at 1 AM, only on Mondays. Used for weekly DBT refreshes.",
)
```

Note that the snapshots are near the end of the DBT DAG. And here's how the runs look:

![image](https://github.com/dagster-io/dagster/assets/52061313/1b30e356-e664-4f62-978a-c89100a0bc70)

One cool thing about this, by the way, is that Dagster figures out that there are some DBT models that are dependent on the snapshots, so even though `dbt_models` has all my models, Dagster still split their materialization into two ops: one for the parents of the snapshots and one for the children of the snapshots!
